// TechpireAPRS/connection/JavAPRSISConnection.js
var util = require('util')
	, net = require('net')
	, EventEmitter = require('events').EventEmitter
	, AbstractDataConnection = require('./AbstractDataConnection.js')
	, APRSPacketParser = require('../util/packets/APRSPacketParser.js')
    , AbstractAPRSMessage = require('../packets/AbstractAPRSMessage.js')
	;

var MESSAGE_DELIMITER = '\r\n';

function JavAPRSISConnection(args) {
	var self = this;

	// TODO: Set defaults
	if(!args) {
		return;
	}
	
	this.callsign = args['callsign'];
	this.passcode = args['passcode'];
	this.filter = args['filter'];
	
	// RESETTING AN INTERVAL EVERY TIME A MESSAGE IS RECEIVED IS EXTREMELY INEFFICIENT.  COME UP WITH SOMETHING BETTER.
	this.keepAliveTime = null;
	
	if(args['keepAliveTime'] != null && args['keepAliveTime'] != undefined && args['keepAliveTime'] > 0) {
		this.keepAliveTime = args['keepAliveTime'];
		this.keepAlive = setInterval(this.SendLogin, this.keepAliveTime, self);
	}
	
	AbstractDataConnection.apply(this, arguments);
	
	return this;
}

JavAPRSISConnection.prototype = new AbstractDataConnection();
JavAPRSISConnection.prototype.constructor = JavAPRSISConnection;

JavAPRSISConnection.prototype.Read = function() {
	// when you get into socket.on, this takes on the socket scope not the JavAPRSISConnection scope
	var self = this;
	
	try {
		var parser = new APRSPacketParser();
		
		this.socket.on('data', function(data) {
			var packets = data.toString().split('\r\n');
			
			for(var x = 0; x < packets.length && packets[x] != ''; x++) {
				if(packets[x].toString()[0] != '#') {
					// parse the packet
					try {
						var rpt = parser.parse(packets[x]);
						
						// finding the message type in a nested if statement seems to be faster than storing the message types in an array to reduce the code
						if(rpt.messageType == '>') {
							// status report
						} else if(rpt.messageType == "$GPGGA,") {
							// GPPA
						} else if(rpt.messageType == '`'
								|| rpt.messageType == '\'') {
							// mic-e
							//console.log(rpt);
							self.emit('position', rpt);
						} else if(rpt.messageType == '='
								|| rpt.messageType == '!'
								|| rpt.messageType == '@'
								|| rpt.messageType == '/') {
							// position report
							//console.log(rpt);
							self.emit('position', rpt);
						} else if(rpt.messageType == '_') {
							// positionless weather report
						} else if(rpt.messageType == 'T') {
							// telemetry data
						} else if(rpt.messageType == ')') {
							//")" Item - bodyBytes.Length > 18
						} else if(rpt.messageType == ';') {
							//";" Object - bodyBytes.Length > 29
							self.emit('object', rpt);
						} else if(rpt.messageType == '}') {
							//"}" Third-party traffic
						} else if(rpt.messageType == ':') {
							// message
							self.emit('message', rpt);
						}
					} catch(e) {
						//console.log('Parse Error: ' + e);
					}
				} else if(packets[x].indexOf("# javAPRSSrvr") == 0 || packets[x].indexOf("# aprsc") == 0) {
					//console.log('received: ' + packets[x]);
					self.SendLogin(self);
					
				}
			}
		});
	} catch(e) {
		throw e;
	}
};

JavAPRSISConnection.prototype.SendLogin = function(instance) {
	var self = instance;

	try {
		//'user <callsign> pass <passcode> vers <softwareName> <softwareVersion> filter <filter>' + MESSAGE_DELIMITER
		
		var serverLogin = 'user ' + self.callsign;
		serverLogin += ' pass ' + self.passcode;
		
		serverLogin += ' vers ' + self.softwareName;
		
		if(self.softwareVersion !== null && self.softwareVersion != undefined && self.softwareVersion !== '') {
			serverLogin += ' ' + self.softwareVersion;
		}
		
		if(self.filter != null && self.filter != '') {
			serverLogin += ' filter ' + self.filter;
		}
		
		serverLogin += MESSAGE_DELIMITER;
		
		self.emit('sending', serverLogin);
		
		self.socket.write(serverLogin, 'utf8');
		
		if(self.keepAliveTime != null && self.keepAliveTime != undefined && self.keepAliveTime > 0) {
			clearInterval(self.keepAlive);
			self.keepAlive = setInterval(self.SendLogin, self.keepAliveTime, self);
		}
	} catch(e) {
		console.log(e);
		self.emit('exception', e);
	}
};

/*
function SendMessage(AbstractAPRSMessage message) {
    // TODO: Implement the ability to send a message
};
*/

module.exports = JavAPRSISConnection;
