// TechpireAPRS/index.js
// Main exports
module.exports.AbstractDataConnection = require('./connection/AbstractDataConnection.js');
module.exports.AbstractDataConnection = require('./connection/AbstractDataConnection.js');
module.exports.JavAPRSISConnection = require('./connection/JavAPRSISConnection.js');
module.exports.APRSDataConnectionFactory = require('./connection/APRSDataConnectionFactory.js');

// Packet type exports
module.exports.AbstractAPRSMessage = require('./packets/AbstractAPRSMessage.js');
module.exports.APRSMessage = require('./packets/APRSMessage.js');
module.exports.APRSPositionReport = require('./packets/APRSPositionReport.js');
module.exports.ObjectReport = require('./packets/ObjectReport.js');

// Parser exports
module.exports.AbstractPacketParser = require('./util/packets/AbstractPacketParser.js');
module.exports.APRSPacketParser = require('./util/packets/APRSPacketParser.js');
module.exports.AGWPEPacketParser = require('./util/packets/AGWPEPacketParser.js');

// Packet Converter exports
module.exports.AbstractPacketConverter = require('./util/packets/AbstractPacketConverter.js');
module.exports.AGWPEPacketConverter = require('./util/packets/AGWPEPacketConverter.js');
module.exports.ISPacketConverter = require('./util/packets/ISPacketConverter.js');
